common/xslt/dhq2html.xsl

```diff
@@ -237,6 +239,171 @@
+                                     name, university, and url-->
+            <xsl:if test="starts-with(., $tabs[$num])">
+                    <!-- Select the chosen matching row -->
+                    <xsl:variable name="inner_tabs" select="tokenize(., '\t')" />
+                    <!-- Editing the url to get the relative url path. index 16 is the url-->
+                    <xsl:variable name="path" select="substring-after($inner_tabs[16], 'org')" />
+                    <!-- Now we pull from the tabs of the selected row. 5 is the name of the article -->
+                    <!-- Setting up the url and its <a> tag-->
+                    <a href="{$path}"><xsl:value-of select="$inner_tabs[5]"/></a>&#160;
+                    <!-- selecting and inputting the year [2], author(s) [3], and university [4] -->
+                    <xsl:value-of select="$inner_tabs[2]"/>,
+                    <xsl:value-of select="$inner_tabs[3]"/>, <xsl:value-of select="$inner_tabs[4] "/>
+            </xsl:if>
+        </xsl:for-each>
+    </xsl:function>
+
+
+    <xsl:template name="recommendations">
+        <!-- Find the row that matches this article's ID. -->
+        <div id="recommendations">
+        <h2>Recommendations</h2>
+        <p>DHQ is testing out three new article recommendation methods! Please explore the links below to find articles that are related in different ways to the one you just read. 
+        We are interested in how these methods work for readers—if you would like to share feedback with us, please complete our short evaluation survey. 
+        You can also visit our <a href="/dhq/explore/explore.html">documentation</a> for these recommendation methods to learn more.</p>
+        <!-- Tokenize rows -->
+        <xsl:variable name="spector_rows" select="tokenize($specter-recs-tsv, '\n')" />
+        <xsl:for-each select="$spector_rows">
+            <xsl:if test="starts-with(., xs:string($id))">
+                <!-- Tokenize that row, splitting by tab character. -->
+                <xsl:variable name="spector_tabs" select="tokenize(., '\t')" />
+                <!-- Set up the container for recommendations: -->
+                    <h3>SPECTER Recommendations</h3>
+                    <p>Below are article recommendations generated by the SPECTER model:</p>
+                    <ol>
+                        <li>
+                            <!-- create variable to pass as parameter. This is index 6 of the tsv, or the first recommended article ID -->
+                            <xsl:variable name="input_num" select="6" />
+                            <!-- Calling the function to populate the recommendation. Pass in the input number, the current row of the selected Article,
+                            and the whole set of rows in the tsv. -->
+                            <xsl:sequence select="dhqf:recs($input_num, $spector_tabs, $spector_rows)" />
+                        </li>
```

Comment:
Here, we could actually use an `<xsl:for-each>` to remove some of the duplicated code in this list. `<xsl:for-each>` doesn't _just_ work on XML nodes or a pre-generated sequence of things. You can also hand it a range of numbers (TSV indexes? 👀) and iterate over each one.

The code below will output 6 messages to the console: 1, 2, 3, 4, 5, 6
```xslt
<xsl:for-each select="1 to 6">
  <xsl:message select="."/>
</xsl:for-each>
```

So, we can replace all those `<li>`s with a single `<xsl:for-each>`:
```xslt
                        <!-- The recommendations for this article start at column 6 of the TSV and end at column 10 -->
                        <xsl:for-each select="6 to 10">
                          <li>
                            <!-- create variable to pass as parameter. This is the index of a recommended article ID -->
                            <xsl:variable name="input_num" select="." />
                            <!-- Calling the function to populate the recommendation. Pass in the input number, the current row of the selected Article,
                            and the whole set of rows in the tsv. -->
                            <xsl:sequence select="dhqf:recs($input_num, $spector_tabs, $spector_rows)" />
                          </li>
                        </xsl:for-each>
```

Given that this code is also repeated for each of the three sections, we could further abstract this code out by making it a function. It'd only need the current row's cells and the other rows:

```xslt
    <xsl:function name="dhqf:make-recs-list-items">
      <xsl:param name="article-row-cells"/>
      <xsl:param name="all-rows"/>
      <!-- The recommendations for this article start at column 6 of the TSV and end at column 10. -->
      <ol>
        <xsl:for-each select="6 to 10">
          <li>
            <!-- create variable to pass as parameter. This is the index of a recommended article ID -->
            <xsl:variable name="input_num" select="." />
            <!-- Calling the function to populate the recommendation. Pass in the input number, the current row of the selected Article,
              and the whole set of rows in the tsv. -->
            <xsl:sequence select="dhqf:recs($input_num, $article-row-cells, $all-rows)" />
          </li>
        </xsl:for-each>
      </ol>
    </xsl:function>
```
We could then call that function for each set of recommendations, e.g.
```xslt
                    <p>Below are article recommendations generated by the SPECTER model:</p>
                    <xsl:sequence select="dhqf:make-recs-list($spector_tabs, $spector_rows)"/>
```

@joelsjlee, I've got a partially-working solution for this too. Would you like me to commit it, save you some time?

Reply:
There's an inconsistency in the code I provided: I changed the function name to `dhqf:make-recs-list()` and used that in the `<xsl:sequence>`, but didn't update the function name in my comment! My bad.

Reply:
Wow again! I should've figured that there was some way to do this in XSLT. Thanks for all the help and yes I would appreciate if you can go ahead and commit the XSLT and CSS changes.

Reply:
On it!